docs/streaming/structured-streaming-transform-with-state.md

```diff
@@ -0,0 +1,324 @@
+        # Calculate downtime duration
+        downtime_duration = timerValues.getCurrentProcessingTimeInMs() - int(time.time() * 1000)
+        # Register a new timer for 10 seconds in the future
+        self.handle.registerTimer(timerValues.getCurrentProcessingTimeInMs() + 10000)
+        # Yield a DataFrame with the key and downtime duration
+        yield pd.DataFrame(
+            {
+                "id": key,
+                "timeValues": str(downtime_duration),
+            }
+        )
+
+    def handleInputRows(self, key, rows, timerValues) -> Iterator[pd.DataFrame]:
+        # Find the row with the maximum timestamp
+        max_row = max((tuple(pdf.iloc[0]) for pdf in rows), key=lambda row: row[1])
+
+        # Get the latest timestamp from existing state or use epoch start if not exists
+        if self.last_seen.exists():
+            latest_from_existing = self.last_seen.get()
+        else:
+            latest_from_existing = datetime.fromtimestamp(0)
+
+        # If new data is more recent than existing state
+        if latest_from_existing < max_row[1]:
+            # Delete all existing timers
+            for timer in self.handle.listTimers():
+                self.handle.deleteTimer(timer)
+            # Update the last seen timestamp
+            self.last_seen.update((max_row[1],))
+
+        # Register a new timer for 5 seconds in the future
+        self.handle.registerTimer(timerValues.getCurrentProcessingTimeInMs() + 5000)
+
+        # Get current processing time in milliseconds
+        timestamp_in_millis = str(timerValues.getCurrentProcessingTimeInMs())
+
+        # Yield a DataFrame with the key and current timestamp
+        yield pd.DataFrame({"id": key, "timeValues": timestamp_in_millis})
+
+    def close(self) -> None:
+        # No cleanup needed
+        pass
+
+{% endhighlight %}
+
+</div>
+
+<div data-lang="scala"  markdown="1">
+
+{% highlight scala %}
+
+// The (String, Timestamp) schema represents an (id, time). We want to do downtime
+// detection on every single unique sensor, where each sensor has a sensor ID.
+class DowntimeDetector(duration: Duration) extends
+  StatefulProcessor[String, (String, Timestamp), (String, Duration)] {
+
+  @transient private var _lastSeen: ValueState[Timestamp] = _
+
+  override def init(outputMode: OutputMode, timeMode: TimeMode): Unit = {
+    _lastSeen = getHandle.getValueState[Timestamp]("lastSeen", Encoders.TIMESTAMP, TTLConfig.NONE)
+  }
+
+  // The logic here is as follows: find the largest timestamp seen so far. Set a timer for
+  // the duration later.
+  override def handleInputRows(
+      key: String,
+      inputRows: Iterator[(String, Timestamp)],
+      timerValues: TimerValues): Iterator[(String, Duration)] = {
+    val latestRecordFromNewRows = inputRows.maxBy(_._2.getTime)
+
+    // Use getOrElse to initiate state variable if it doesn't exist
+    val latestTimestampFromExistingRows = _lastSeen.getOption().getOrElse(new Timestamp(0))
```

Comment:
getOption() was removed

Reply:
Why not double confirm the code we put here with the artifact of current branch-4.0?

Reply:
Done